pkg/operator/ceph/spec/spec.go

```diff
@@ -363,15 +363,28 @@ func generateLifeCycleCmd(dataDirHostPath string) []string {
 	return cmd
 }
 
-// PodLifeCycle returns a pod lifecycle resource to execute actions before a pod starts
-func PodLifeCycle(dataDirHostPath string) *v1.Lifecycle {
+// PodLifeCycle returns a pod lifecycle resource to execute actions before a pod starts and before a pod terminates
+func PodLifeCycle(dataDirHostPath, volumeGroup string) *v1.Lifecycle {
 	cmd := generateLifeCycleCmd(dataDirHostPath)
 
-	return &v1.Lifecycle{
+	lifeCyleHooks := &v1.Lifecycle{
 		PostStart: &v1.Handler{
 			Exec: &v1.ExecAction{
 				Command: cmd,
 			},
 		},
 	}
+
+	// release device from lvm
+	if volumeGroup != "" {
+		lifeCyleHooks.PreStop = &v1.Handler{
+			Exec: &v1.ExecAction{
+				Command: []string{"/sbin/lvchange", "-an", volumeGroup},
```

Comment:
- Have you found scenarios where the preStop hook does not execute? For example, what if the ceph-osd process crashes? This could be simulated by exec'ing to the pod and killing the ceph-osd process. Should we call this `lvchange -an` command after the process exits? 
- Is `lvchange -an` a no-op if not needed? For example, could we call this command [here](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/osd/daemon.go#L73) after the `ceph-osd` process exits, as another place to possibly free the LV.

Reply:
@nixpanic : Thoughts on PreStop hook killing the ceph-osd process before deactivating the LV. Something like this
```
cmd := []string{
		"/bin/sh",
		"-c",
		"killall ceph-osd",
		fmt.Sprintf("/sbin/lvchange -an %s", vgGroup),
	}
```

PS: haven't tested this out!

Reply:
Something like that might be better. I'd use `killall --wait ...` though. In case the osd process needs some time to cleanly exit and close the device.

Note that you may need to pass the `killall` and `lvchange` commands as a single string with `;` as seperator.

Reply:
nit: I think simply calling `lvchange` instead of the full path should be enough.

Reply:
so I updated the preStop hook and to kill the  ceph-osd process. The OSD pod logs are below 

```
2019-09-04 15:04:59.758792 I | Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
2019-09-04 15:05:00.374247 I | --> ceph-volume lvm activate successful for osd ID: 0
2019-09-04 15:05:00.382401 I | exec: Running command: ceph-osd --foreground --id 0 --osd-uuid 3ee0ffd8-db80-4631-a50c-12b7b9ccef37 --conf /var/lib/rook/osd0/rook-ceph.config --cluster ceph --setuser ceph --setgroup ceph --setuser-match-path /var/lib/rook/osd0 --default-log-to-file false --ms-learn-addr-from-peer=false
2019-09-04 15:05:00.739720 I | 2019-09-04 15:05:00.738 7fc1ea811d80 -1 Falling back to public interface
2019-09-04 15:05:01.299658 I | 2019-09-04 15:05:01.298 7fc1ea811d80 -1 osd.0 0 log_to_monitors {default=true}
2019-09-04 15:05:03.284168 I | 2019-09-04 15:05:03.283 7fc1d4ba1700 -1 osd.0 0 waiting for initial osdmap
2019-09-04 15:05:03.290203 I | 2019-09-04 15:05:03.289 7fc1dce1c700 -1 osd.0 4 set_numa_affinity unable to identify public interface 'eth0' numa node: (2) No such file or directory
2019-09-04 15:14:52.987616 I | 2019-09-04 15:14:52.985 7fc1e06ae700 -1 received  signal: Terminated from killall ceph-osd  (PID: 113997) UID: 0
2019-09-04 15:14:52.987671 I | 2019-09-04 15:14:52.985 7fc1e06ae700 -1 osd.0 5 *** Got signal Terminated ***
2019-09-04 15:14:54.067639 I | exec: Running command: /sbin/lvchange -an ceph-d9765f67-f5e1-4d70-8d52-e9bbab9b5117
2019-09-04 15:14:54.094943 I |   WARNING: Not using device /dev/xvdbo for PV rbRs6e-CpqG-LS8j-fdqC-dahn-dEUP-51uNAN.
2019-09-04 15:14:54.095017 I |   WARNING: PV rbRs6e-CpqG-LS8j-fdqC-dahn-dEUP-51uNAN prefers device /dev/loop0 because device is used by LV.
2019-09-04 15:14:54.128123 I | op-osd: Successfully released device from lvm
rpc error: code = Unknown desc = container with ID starting with a0a83e7decfc2a25ed55937731b1c2b54ca250b56c34dd134d53e9dda4e11a87 not found: ID does not exist[sapillai@localhost ceph]$ 
```

I see the killall  lvchange command being run when I delete the rook cluster. But still PV remains in failed state with the error: 

`Error deleting EBS volume "vol-0a256dfc3141b62ba" since volume is currently attached to "i-0a8d04dbdca45942c"`

Reply:
Should the command not be `/sbin/lvchange -an <VGname>/ceph-d9765f67-f5e1-4d70-8d52-e9bbab9b5117`? Capturing and logging the output of `lvchange` would be useful. Maybe run it with `-vvv` while testing.

Reply:
You may even need to run the `lvchange` command through `nsenter` so that the command gets executed on the host (and not inside the container).

Reply:
> Should the command not be `/sbin/lvchange -an <VGname>/ceph-d9765f67-f5e1-4d70-8d52-e9bbab9b5117`? Capturing and logging the output of `lvchange` would be useful. Maybe run it with `-vvv` while testing.

`ceph-c8817873-b005-44a1-8fb1-79362a762491` is the VGname here. 

```
sh-4.4# lvs
  WARNING: Not using device /dev/xvdcg for PV EzdF2V-q9Rb-bVPD-JTgE-lYea-0yeL-afWc1F.
  WARNING: PV EzdF2V-q9Rb-bVPD-JTgE-lYea-0yeL-afWc1F prefers device /dev/loop0 because device is used by LV.
  LV                                             VG                                        Attr       LSize Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  osd-block-da2069ae-9eff-40b4-a736-2fe6c3691a3a ceph-c8817873-b005-44a1-8fb1-79362a762491 -wi-ao---- 9.00g   
```